Add docs for execute, prove and verify, update docs (#2036)
software-mansion/scarb: website/docs/extensions/execute.md

<script setup>
import { data as rel } from "../../github.data";
</script>

# Scarb execute

The `scarb execute` command executes a function from a local package.
Only packages defining the [executable target](../reference/targets#executable-target) can be executed.
It does automatically compile the cairo code within the package so using `scarb build` beforehand is not necessary.

> This automatically called build can be skipped with the `--no-build` flag.

Author: I don't want to push it too hard, if they are bothered with the build time, they will figure it out 🤔

> This automatically called build can be skipped with the `--no-build` flag.

Reviewer: I don't insist on pushing it too hard, but it's not explicitly mentioned under which conditions it's meaningful to use `--no-build`. Maybe we could have something like:

```
If `scarb build` or `scarb execute` has been previously used and the package hasn't changed since, this automatic build can be optionally skipped with the `--no-build` flag.
```

wdyt?

## Choosing a function to run

If your package defines multiple main functions (see [choosing the main function](../reference/targets#choosing-the-main-function)),
you need to specify which executable target should be run.

This can be achieved through one of two flags:

- `--executable-name` to choose the target by its name.
- `--executable-function` to choose the target by the main function it defines.
  Those flags are mutually exclusive.

## Saving the execution information

The execution will be carried out for one of two execution targets: `standalone` or `bootloader`.
You can choose the target with the `--target` flag.
Standalone means that the program will be executed as-is, and intended to be proven directly with `scarb prove`.
hen we run with the bootloader target, the program’s execution is expected to be wrapped by the
[bootloader’s](https://github.com/Moonsong-Labs/cairo-bootloader?tab=readme-ov-file#cairo-bootloader) execution,
which itself will be proven via Stwo.

> After the execution, information about it will be saved to the target directory of the package.
> 
> For `standalone` target, the output will be saved as trace files (`air_public_input.json`, `air_private_input.json`,
> `memory.bin`, and `trace.bin`), which can be used for creating a proof with `scarb prove`.
> For `bootloader` target, the output will be saved as a CairoPie format (Position Indenpendent Execution),
> which is not yet supported by the `prove` command.

Reviewer: as above
```suggestion
After the execution, information about it will be saved to the target directory of the package:

- For `standalone` target, the output will be saved as trace files (`air_public_input.json`, `air_private_input.json`,
  `memory.bin`, and `trace.bin`), which can be used for creating a proof with `scarb prove`.
- For `bootloader` target, the output will be saved as a CairoPie format (Position Indenpendent Execution),
  which is not yet supported by the `scarb prove` command.
```

This will show information about:

- `n_steps`
- `n_memory_holes`
- `builtin_instance_counter`
- `syscalls`

## Program arguments

The executable function may accept arguments.
They can be passed to the `scarb execute` command via either `--arguments` or `--arguments-file` flag.

The expected input with `--arguments` is a comma-separated list of integers.
This list should correspond to the Cairo’s Serde of main’s arguments, for example:

| main’s signature                       | valid arguments example |
| :------------------------------------- | :---------------------- |
| `fn main(num: u8)`                     | 1                       |
| `fn main(num1: u8, num2: u16)`         | 1,2                     |
| `fn main(num1: u8, tuple: (u16, u16))` | 1,2,3                   |
| `fn main(num1: u8, num2: u256)`        | 1,2,3                   |
| `fn main(num1: u8, arr: Array<u8>)`    | 1,2,1,2                 |

See the [documentation](https://docs.starknet.io/architecture-and-concepts/smart-contracts/serialization-of-cairo-types/) for more information about Cairo’s Serde.

Note that when using `--arguments-file`, the expected input is an array of felts represented as hex string.

> For example, `1,2,3` in the above table becomes `["0x1", "0x2", "0x3"]`.

Reviewer: Hmmm, maybe we could just update the table above to have:
- valid `--arguments` example
- valid `--arguments-file` example

Would be mildly useful to add one number like `27` == `0x1b`

<script setup>
import { data as rel } from "../../github.data";
</script>

# Scarb execute

> The `scarb execute` command executes a function from a local package.
> Only packages defining the [executable target](../reference/targets#executable-target) can be executed.
> It does automatically compile the Cairo code within the package so using `scarb build` beforehand is not necessary.
> This automatically called build can be skipped with the `--no-build` flag.

Reviewer 2: ```suggestion
The `scarb execute` command executes a function from a local package.
It does automatically compile the Cairo code within the package so using `scarb build` beforehand is not necessary.
You can skip the automatic build with the `--no-build` flag.

Only packages defining the [executable target](../reference/targets#executable-target) can be executed.
```
Would be a bit clearer imo, but no strong feelings.

This will show information about:

- `n_steps`
- `n_memory_holes`
- `builtin_instance_counter`
- `syscalls`

## Program arguments

The executable function may accept arguments.
They can be passed to the `scarb execute` command via either `--arguments` or `--arguments-file` flag.

The expected input with `--arguments` is a comma-separated list of integers.
This list should correspond to the Cairo’s Serde of main’s arguments, for example:

| main’s signature                       | valid arguments example |
| :------------------------------------- | :---------------------- |
| `fn main(num: u8)`                     | 1                       |
| `fn main(num1: u8, num2: u16)`         | 1,2                     |
| `fn main(num1: u8, tuple: (u16, u16))` | 1,2,3                   |
| `fn main(num1: u8, num2: u256)`        | 1,2,3                   |
| `fn main(num1: u8, arr: Array<u8>)`    | 1,2,1,2                 |

See the [documentation](https://docs.starknet.io/architecture-and-concepts/smart-contracts/serialization-of-cairo-types/) for more information about Cairo’s Serde.

Note that when using `--arguments-file`, the expected input is an array of felts represented as hex string.

> For example, `1,2,3` in the above table becomes `["0x1", "0x2", "0x3"]`.

Reviewer 2: I think we could have a separate section for command's flags. Something like for example [pip](https://pip.pypa.io/en/stable/cli/pip_freeze/#options:~:text=insensitive%20sorted%20order.-,Options,-%C2%B6). 
Please consider (not necessarily for this PR but generally).

Author: We could do that if we had some way to auto-generate those. Otherwise, I wouldn't really want to burden us with listing all flags and keeping them up to date 🤔
